server-common/src/main/java/org/apache/kafka/server/common/FeatureVersion.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.server.common;
+
+import java.util.Map;
+
+public interface FeatureVersion {
+
+    /**
+     * The level of the feature. 0 means the feature is disabled.
+     */
+    short featureLevel();
+
+    /**
+     * The name of the feature.
+     */
+    String featureName();
+
+    /**
+     * The minimum metadata version which sets this feature version as default. When bootstrapping using only
+     * a metadata version, a reasonable default for all other features is chosen based on this value.
+     * This should be defined as the next metadata version to be released when the feature version becomes production ready.
+     * (Ie, if the current production MV is 17 when a feature version is released, its mapping should be to MV 18)
```

Comment:
Can we make it map to the release version?  If not, can we add a comment explaining the logic?

Intuitively, if we have a feature like
- FOO(1, 42) // released version 1 when MV got bumped to 42 
- FOO(2, 77) // released version 2 when MV got bumped to 77
- FOO(3, 77) // released version 3 at the same time as 2

we should be able to discover that if we got MV 45, then FOO=1 because 1 was available at 42 and 2&3 are not available yet.  If we got 78 then it should be 3 because it was max available at 77.

If we got MV 30 then we should get 0 (implicit) because it wasn't available yet.

Using the same table, we can determine that if a feature version 1 was selected with MV 41, then it's invalid combination; similarly if feature 2 or 3 was selected with MV 45 it's invalid combination.  And we can select 1, 2 or 3 when MV is 77.

Reply:
How do we know the release version when we create the feature. The release version can change, so should we change it as each new MV is added?

This is also what I did but adding 1 to every MV for the reasons I explained [here](https://github.com/apache/kafka/pull/15685#discussion_r1610248735)

Reply:
> How do we know the release version when we create the feature.

My understanding is that the new version of the feature will start as an unstable one. When we promote it to production ready, we can attach it to the correct MV (the latest one available in the release, I suppose).

Reply:
My thought process is that we will just attach it to the current next unstable one. If there is no such MV we create it and if there is still no feature atttached to it, we will just release it at the release. I probably need to write up a formal doc on how this will work.

Reply:
Understood. The downside of this is that you have to ensure that you change it when the next unstable one is released and the new feature is not ready to be promoted.

Reply:
Sorry I think I'm not explaining this clearly. Once it is released, this value shouldn't change. I wish KIP-1014 was completed so this could be clearer.